vscode/src/projectSystem.ts

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { log } from "qsharp-lang";
+import { Utils, URI } from "vscode-uri";
+import * as vscode from "vscode";
+
+// This flag is a hacky way of disabling the debugger when running in project mode.
+// We will have debugger support for projects _very soon_, and this flag will get
+// removed when that lands. To be clear, this is _not_ the ideal way to transmit
+// data to the debugger, but it is a temporary workaround.
+// It stores files that are a part of projects and disables the debugger on them.
+let PROJECT_MODE = false;
+export const getProjectMode = () => PROJECT_MODE;
+
+/**
+ * Finds and parses a manifest. Returns `null` if no manifest was found for the given uri, or if the manifest
+ * was malformed.
+ */
+export async function getManifest(uri: string): Promise<{
+  excludeFiles: string[];
+  excludeRegexes: string[];
+  manifestDirectory: string;
+} | null> {
+  const manifestDocument = await findManifestDocument(uri);
+
+  if (manifestDocument === null) {
+    PROJECT_MODE = false;
+    return null;
+  }
+
+  let parsedManifest;
+  try {
+    parsedManifest = JSON.parse(manifestDocument.content);
+  } catch (e) {
+    log.error(
+      "Found manifest document, but the Q# manifest was not valid JSON",
+      e,
+    );
+    PROJECT_MODE = false;
+    return null;
+  }
+
+  const manifestDirectory = Utils.dirname(manifestDocument.uri);
+
+  PROJECT_MODE = true;
+  return {
+    excludeFiles: parsedManifest.excludeFiles || [],
+    excludeRegexes: parsedManifest.excludeRegexes || [],
+    manifestDirectory: manifestDirectory.toString(),
+  };
+}
+
+/** Returns the manifest document if one is found
+ * returns null otherwise
+ */
+async function findManifestDocument(
+  currentDocumentUriString: string,
+): Promise<{ uri: vscode.Uri; content: string } | null> {
+  // /home/foo/bar/document.qs
```

Comment:
It may be worth noting here some of the more complex URIs you'll need to deal with beyond just a file system path. For example, in github.dev when looking at files that URI can be something like `vscode-vfs://github%2B7b2276223a312c22726566223a7b2274797065223a332c226964223a22383439227d7d/microsoft/qsharp/samples/shor.qs`. The scheme, authority, and path are all critical to maintain in future queries.

Reply:
And for example in our https://vscode.dev/quantum/playground/ experience, where document URIs are like `qsharp-vfs://playground/samples/Entanglement.qs`), which is our own scheme/file-system ("qsharp-vfs") and authority ("playground") before the path ("/samples/Entanglement.qs").